// @miro-site/enterprise-workspaces/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var GLOBAL_WORKSPACE_DOMAIN_NAME = 'app';
var VISITED_WORKSPACES_COOKIE = 'workspaces';
var WORKSPACE_QUERY_PARAMETER = 's';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var fetchUsedWorkspaces = function fetchUsedWorkspaces(_a) {
  var applicationBaseHost = _a.applicationBaseHost,
    fetchWorkspaceInfo = _a.fetchWorkspaceInfo,
    workspacesHostNames = _a.workspacesHostNames;
  return __awaiter(void 0, void 0, void 0, function () {
    var fetchResults;
    return __generator(this, function (_b) {
      switch (_b.label) {
        case 0:
          return [4 /*yield*/, Promise.allSettled(workspacesHostNames.map(function (workspaceHostName) {
            return fetchWorkspaceInfo(workspaceHostName);
          }))];
        case 1:
          fetchResults = _b.sent();
          return [2 /*return*/, fetchResults.reduce(function (acc, result) {
            if (result.status === 'fulfilled') {
              acc[getWorkspaceHostName(result.value, applicationBaseHost)] = result.value;
            }
            return acc;
          }, {})];
      }
    });
  });
};
var sortRegionTypes = function sortRegionTypes(a, b) {
  if (a.regionType === b.regionType) {
    return 0;
  }
  if (a.regionType === 'eu' && b.regionType !== 'eu' || a.regionType !== 'private' && b.regionType === 'private') {
    return -1;
  }
  if (a.regionType !== 'eu' && b.regionType === 'eu' || a.regionType === 'private' && b.regionType !== 'private') {
    return 1;
  }
  return a.regionType.localeCompare(b.regionType);
};
var sanitizeWorkspaceBaseHost = function sanitizeWorkspaceBaseHost(baseHost, applicationBaseHost) {
  var baseHostParts = baseHost.split('.');
  if (baseHostParts.length < 2 || baseHost === applicationBaseHost) {
    return applicationBaseHost;
  }
  var applicationBaseHostParts = applicationBaseHost.split('.');
  if (baseHost.endsWith(".".concat(applicationBaseHost)) && baseHostParts.length - applicationBaseHostParts.length === 1) {
    return [baseHostParts[0], applicationBaseHost].filter(function (part) {
      return !!part;
    }).join('.');
  }
  return applicationBaseHost;
};

var getWorkspaceHostName = function getWorkspaceHostName(_a, applicationBaseHost) {
  var workspaceBaseHost = _a.baseHost,
    domainName = _a.domainName;
  var sanitizedWorkspaceBaseHost = sanitizeWorkspaceBaseHost(workspaceBaseHost || '', applicationBaseHost);
  if (!domainName || domainName === sanitizedWorkspaceBaseHost) {
    return sanitizedWorkspaceBaseHost;
  }
  var domainNameParts = domainName.split('.');
  if (domainNameParts.length === 1) {
    return [domainNameParts[0], sanitizedWorkspaceBaseHost].filter(function (part) {
      return !!part;
    }).join('.');
  }
  var sanitizedWorkspaceBaseHostParts = sanitizedWorkspaceBaseHost.split('.');
  if (domainName.endsWith(".".concat(sanitizedWorkspaceBaseHost)) && domainNameParts.length - sanitizedWorkspaceBaseHostParts.length === 1) {
    return [domainNameParts[0], sanitizedWorkspaceBaseHost].filter(function (part) {
      return !!part;
    }).join('.');
  }
  return sanitizedWorkspaceBaseHost;
};
var getVisitedWorkspacesHostNames = function getVisitedWorkspacesHostNames(cookieValue, applicationBaseHost) {
  return cookieValue ? cookieValue.split(',').map(function (workspaceDomain) {
    return getWorkspaceHostName({
      domainName: workspaceDomain === GLOBAL_WORKSPACE_DOMAIN_NAME ? '' : workspaceDomain
    }, applicationBaseHost);
  }) : [];
};
var getVisitedWorkspacesCookieAttributes = function getVisitedWorkspacesCookieAttributes(applicationBaseHost, isDelete) {
  return {
    domain: applicationBaseHost,
    httpOnly: false,
    maxAge: isDelete ? 0 : 31536000,
    path: '/',
    sameSite: 'None',
    secure: true
  };
};
var addVisitedWorkspace = function addVisitedWorkspace(currentCookieValue, workspace, applicationBaseHost) {
  var value = new Set(getVisitedWorkspacesHostNames(currentCookieValue, applicationBaseHost));
  value.add(getWorkspaceHostName(workspace, applicationBaseHost));
  return {
    attributes: getVisitedWorkspacesCookieAttributes(applicationBaseHost),
    name: VISITED_WORKSPACES_COOKIE,
    value: Array.from(value).join(',')
  };
};
var removeVisitedWorkspace = function removeVisitedWorkspace(currentCookieValue, workspace, applicationBaseHost) {
  var value = new Set(getVisitedWorkspacesHostNames(currentCookieValue, applicationBaseHost));
  value["delete"](getWorkspaceHostName(workspace, applicationBaseHost));
  if (value.size === 0) {
    return {
      attributes: getVisitedWorkspacesCookieAttributes(applicationBaseHost, true),
      name: VISITED_WORKSPACES_COOKIE,
      value: ''
    };
  }
  return {
    attributes: getVisitedWorkspacesCookieAttributes(applicationBaseHost),
    name: VISITED_WORKSPACES_COOKIE,
    value: Array.from(value).join(',')
  };
};
var isPrivateWorkspace = function isPrivateWorkspace(_a) {
  var global = _a.global;
  return !global;
};
var isCommonWorkspace = function isCommonWorkspace(workspace, applicationBaseHost) {
  if (isPrivateWorkspace(workspace)) {
    return false;
  }
  return getWorkspaceHostName(workspace, applicationBaseHost) === applicationBaseHost;
};
var isRegionalWorkspace = function isRegionalWorkspace(workspace, applicationBaseHost) {
  if (isPrivateWorkspace(workspace)) {
    return false;
  }
  return getWorkspaceHostName(workspace, applicationBaseHost) !== applicationBaseHost;
};
var getRegionType = function getRegionType(_a) {
  var applicationBaseHost = _a.applicationBaseHost,
    workspaceHostName = _a.workspaceHostName;
  if (applicationBaseHost === workspaceHostName) {
    return 'eu';
  }
  var region = workspaceHostName.replace(".".concat(applicationBaseHost), '');
  switch (region) {
    case 'au':
    case 'jp':
    case 'us':
      return region;
    default:
      return 'private';
  }
};
var getRememberedProfileWorkspaceHostNames = function getRememberedProfileWorkspaceHostNames(_a) {
  var applicationBaseHost = _a.applicationBaseHost,
    rememberedProfiles = _a.rememberedProfiles;
  var workspaceHostNames = new Set();
  rememberedProfiles.forEach(function (profile) {
    workspaceHostNames.add(getWorkspaceHostName(profile.workspace || {
      baseHost: applicationBaseHost,
      domainName: ''
    }, applicationBaseHost));
  });
  return workspaceHostNames;
};
var groupRememberedProfilesByWorkspaceType = function groupRememberedProfilesByWorkspaceType(_a) {
  var applicationBaseHost = _a.applicationBaseHost,
    fetchWorkspaceInfo = _a.fetchWorkspaceInfo,
    rememberedProfiles = _a.rememberedProfiles;
  return __awaiter(void 0, void 0, void 0, function () {
    var workspacesMap;
    return __generator(this, function (_b) {
      switch (_b.label) {
        case 0:
          return [4 /*yield*/, fetchUsedWorkspaces({
            applicationBaseHost: applicationBaseHost,
            fetchWorkspaceInfo: fetchWorkspaceInfo,
            workspacesHostNames: Array.from(getRememberedProfileWorkspaceHostNames({
              applicationBaseHost: applicationBaseHost,
              rememberedProfiles: rememberedProfiles
            }))
          })];
        case 1:
          workspacesMap = _b.sent();
          return [2 /*return*/, Object.values(rememberedProfiles.reduce(function (acc, profile) {
            var workspaceHostName = getWorkspaceHostName(profile.workspace || {
              baseHost: applicationBaseHost,
              domainName: ''
            }, applicationBaseHost);
            if (workspacesMap[workspaceHostName]) {
              var regionType = getRegionType({
                applicationBaseHost: applicationBaseHost,
                workspaceHostName: workspaceHostName
              });
              if (!acc[regionType]) {
                acc[regionType] = {
                  regionType: regionType,
                  rememberedProfiles: []
                };
              }
              acc[regionType].rememberedProfiles.push(__assign(__assign({}, profile), {
                workspace: workspacesMap[workspaceHostName]
              }));
            }
            return acc;
          }, {})).sort(sortRegionTypes)];
      }
    });
  });
};

exports.GLOBAL_WORKSPACE_DOMAIN_NAME = GLOBAL_WORKSPACE_DOMAIN_NAME;
exports.VISITED_WORKSPACES_COOKIE = VISITED_WORKSPACES_COOKIE;
exports.WORKSPACE_QUERY_PARAMETER = WORKSPACE_QUERY_PARAMETER;
exports.addVisitedWorkspace = addVisitedWorkspace;
exports.getRegionType = getRegionType;
exports.getRememberedProfileWorkspaceHostNames = getRememberedProfileWorkspaceHostNames;
exports.getVisitedWorkspacesCookieAttributes = getVisitedWorkspacesCookieAttributes;
exports.getVisitedWorkspacesHostNames = getVisitedWorkspacesHostNames;
exports.getWorkspaceHostName = getWorkspaceHostName;
exports.groupRememberedProfilesByWorkspaceType = groupRememberedProfilesByWorkspaceType;
exports.isCommonWorkspace = isCommonWorkspace;
exports.isPrivateWorkspace = isPrivateWorkspace;
exports.isRegionalWorkspace = isRegionalWorkspace;
exports.removeVisitedWorkspace = removeVisitedWorkspace;
